jax/_src/dispatch.py

```diff
@@ -283,25 +286,8 @@ def lower_xla_callable(fun: lu.WrappedFun, device, backend, name,
     in_type = tuple(unsafe_zip(abstract_args, itertools.repeat(True)))
     fun = lu.annotate(fun, in_type)
   else:
-    # Check that the provided abstract_args are consistent with in_type by first
-    # collecting values of axis size arguments, then substituting them in for
-    # DBIdx occurrences.
-    axis_sizes: Dict[core.DBIdx, int] = {}
-    abstract_args_iter = iter(abstract_args)
-    for expected_type, explicit in fun.in_type:
-      if explicit:
-        aval = next(abstract_args_iter)
-        if isinstance(expected_type, core.DShapedArray):
-          # Check the value for any DBIdx variables is consistent.
-          assert all(axis_sizes.setdefault(d1, d2) == d2
-                     for d1, d2 in zip(expected_type.shape, aval.shape)
-                     if type(d1) is core.DBIdx)
-          # Check the type matches after substitution.
-          expected_shape = [axis_sizes.get(d, d) for d in expected_type.shape]  # type: ignore
-          expected_aval = core.ShapedArray(
-              shape=tuple(expected_shape), dtype=expected_type.dtype,
-              weak_type=expected_type.weak_type)
-          assert core.typematch(expected_aval, aval)
+    assert abstract_args == (None,) * len(abstract_args)
+    abstract_args = [aval for aval, _ in fun.in_type]
```

Comment:
The previous code was mistaken because it caused cache misses. That is, the previous code correctly compared the arguments' avals (`abstract_args`, unzipped from `arg_specs`) to the type annotation on the function (`fun.in_type`), but if we pass those `avals` as part of `arg_specs` then we'll get the static shapes of arguments as part of the cache key (i.e. affecting the memoization decorator on this function). So instead when the input type is annotated, and hence we don't need to infer the type from the arguments, we don't pass in the arguments' types (see above, around line 195).